File: notificacao/src/main/java/com/logistica/notificacao/model/Notificacao.java
package com.logistica.notificacao.model;

import io.hypersistence.utils.hibernate.type.json.JsonBinaryType;
import jakarta.persistence.*;
import lombok.Getter;
import lombok.Setter;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.annotations.Type;
import org.hibernate.type.SqlTypes;

import java.time.LocalDateTime;
import java.util.Map;

@Entity
@Table(name = "notificacoes")
@Getter
@Setter
public class Notificacao {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    private String tipoEvento;
    private String origem;

    private Long destinatarioId;

    private String titulo;
    private String mensagem;

    private LocalDateTime dataCriacao;
    private LocalDateTime dataLeitura;

    @Enumerated(EnumType.STRING)
    private StatusNotificacao status;

    @Type(JsonBinaryType.class)
    @JdbcTypeCode(SqlTypes.JSON)
    @Column(columnDefinition = "jsonb")
    private Map<String, Object> dadosEvento;

    public enum StatusNotificacao {
        NAO_LIDA, LIDA
    }

    public Map<String, Object> getDadosEvento() {
        return dadosEvento;
    }

    public void setDadosEvento(Map<String, Object> dadosEvento) {
        this.dadosEvento = dadosEvento;
    }
}

File: app/lib/models/Notificacao.dart
import 'dart:convert';

class Notificacao {
  final int id;
  final String tipoEvento;
  final String origem;
  final int destinatarioId;
  final String titulo;
  final String mensagem;
  final String dataCriacao;
  final String? dataLeitura;
  bool lida;
  final String? payload;

  Notificacao({
    required this.id,
    required this.tipoEvento,
    required this.origem,
    required this.destinatarioId,
    required this.titulo,
    required this.mensagem,
    required this.dataCriacao,
    this.dataLeitura,
    this.lida = false,
    this.payload,
  });

  factory Notificacao.fromJson(Map<String, dynamic> json) {
    String? payload = jsonEncode(json);

    return Notificacao(
      id: json['id'],
      tipoEvento: json['tipoEvento'] ?? '',
      origem: json['origem'] ?? '',
      destinatarioId: json['destinatarioId'],
      titulo: json['titulo'] ?? 'Notificação',
      mensagem: json['mensagem'] ?? '',
      dataCriacao: json['dataCriacao'] ?? DateTime.now().toIso8601String(),
      dataLeitura: json['dataLeitura'],
      lida: json['dataLeitura'] != null,
      payload: payload,
    );
  }
}
